### PR TITLE
Read target architecture from cross-compiled binaries.

### LINKED ISSUE #821: Read Target Architecture Information in XTask
Read the target architecture, as if via `readelf -A` for a compiled binary from the target, without producing any build artifacts on the host, and print the output as JSON so that the target info can be verified. The `Tag_` prefixes should also be removed. This could be useful to prevent issues like #426 when adding new targets.

The output for `arm-unknown-linux-gnueabihf` would therefore look like:

```json
{
    "CPU_name": "6",
    "CPU_arch": "v6",
    "ARM_ISA_use": "Yes",
    "THUMB_ISA_use": "Thumb-1",
    "FP_arch": "VFPv2",
    "ABI_PCS_GOT_use": "GOT-indirect",
    "ABI_PCS_wchar_t": "4",
    "ABI_FP_rounding": "Needed",
    "ABI_FP_denormal": "Needed",
    "ABI_FP_exceptions": "Needed",
    "ABI_FP_number_model": "IEEE 754",
    "ABI_align_needed": "8-byte",
    "ABI_enum_size": "int",
    "ABI_VFP_args": "VFP registers",
    "CPU_unaligned_access": "v6",
    "ABI_FP_16bit_format": "IEEE 754"
}
```

### PR DESCRIPTION
**Pull request by Alexhuszagh**
Use `readelf -A` on generated binaries to read the target architecture information and output the data as JSON, simplifying verifying that new targets produce code for the desired architecture.

Closes #821.
Related to #426.

### CODE REVIEW

**Comment by Alexhuszagh** [comment]
I also removed the `-it` flag from `target-info`, since this can cause a wait status to generate, and it isn't necessary for the command to function.

File: xtask/src/readelf.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    command.args(&["bash", "-c", READELF_SCRIPT]);
+
+    if !verbose {
+        // capture stderr to avoid polluting table
+        command.stderr(Stdio::null());
+    }
+    let stdout = command.run_and_get_stdout(verbose)?;
+    let trimmed = stdout.trim();
+    if trimmed.is_empty() {
+        eprintln!("no target information: target is likely same architecture as host.");
```
Comment (Emilgardis) [inline xtask/src/readelf.rs:68]:
```suggestion
    let out = command.run_and_get_output(verbose)?;
    let stdout = out.stdout()?;
    if stdout.trim().is_empty() {
        let stderr = out.stderr()?;
        if !stderr.trim().is_empty() {
            if stderr.contains("target may not be installed") {
                eprintln!("warning: target {target} is not installed");
                return Ok(());
            } else if !stderr.contains("Finished dev") {
                eprintln!("{}", stderr.header("stderr"));
                return Ok(());
            }
        }
        eprintln!("no target information: target `{target}` is likely same architecture as host.");
        if verbose {
            eprintln!("{}", stderr.header("stderr"));
        }
```

File: xtask/src/readelf.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+main() {
+    local td
+    td="$(mktemp -d)"
+
+    pushd "${td}" >/dev/null
+    cargo init --bin hello >/dev/null
+    cd hello
+    cargo build --target "${TARGET}" "${@}" >/dev/null
+    readelf -A "target/${TARGET}/debug/hello"
```
Comment (Emilgardis) [inline xtask/src/readelf.sh:15]:
```suggestion
    cargo init --bin hello 1>&2
    cd hello
    cargo build --target "${TARGET}" "${@}" 1>&2
    readelf -A "target/${TARGET}/debug/hello"
```

File: xtask/src/readelf.rs (hunk 1 above)
Comment (Emilgardis) [reply xtask/src/readelf.rs:68]: uses `color_eyre::SectionExt` to output it nicely

File: xtask/src/util.rs [hunk 3]
```diff
@@ -68,6 +72,30 @@ pub fn get_matrix() -> &'static Vec<Matrix> {
...
+pub fn pull_image(engine: &Path, image: &str, verbose: bool) -> cross::Result<()> {
+    let mut command = Command::new(engine);
+    command.arg("pull");
+    command.arg(image);
+    if !verbose {
+        // capture output to avoid polluting table
+        command.stdout(Stdio::null());
+        command.stderr(Stdio::null());
+    }
+    command.run(verbose, false).map_err(Into::into)
```
Comment (Emilgardis) [inline xtask/src/util.rs:96]: I know this is old but I think this should be changed to

```suggestion
    let out = command.run_and_get_output(verbose)?;
    command.status_result(verbose, out.status, Some(&out))?;
    Ok(())
```

**Review by Emilgardis — CHANGES_REQUESTED** [review]
looks good!

why not run `readelf` on the host?

**Comment by Alexhuszagh** [comment]
This actually needs to install the target architecture via rustup if it's not present.

**Comment by Emilgardis** [comment]
yup, I don't know if maybe it makes sense to do it only when wanted, eg `--download-targets` or something

**Comment by Alexhuszagh** [comment]
Actually this might be more difficult than I thought. MIPS is giving me some very weird data. Probably will close this. You can submit the other changes as a separate PR.

**Pull request closed without merging**